feat: rename `ShellOtps` to `GlobalOpts`

zerosnacks opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.
-->

## Motivation

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

Related: https://github.com/foundry-rs/foundry/issues/8408

## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

- Renames `ShellOpts` to `GlobalOpts`.
- Only selectively passes down `GlobalOpts` where required, this can be the case when wanting to indicate a `conflicts_with` a global argument. In most cases when checking against the shell configuration it is probably easier to use the e.g. `shell::is_json` variant over passing it all the way down.

grandizzy reviewed (changes requested):
impl looks good! there's a regression re the way rayon thread pool is spawned though, a command like
```
forge test --show-progress --jobs 3
```
always fail now with `Error: The global thread pool has already been initialized.`

while the cast create2 is operating on single thread now instead all available
```
cast create2 --starts-with dead --case-sensitive --deployer 0x0000000000FFe8B47B3e2130213B802212439497 --init-code-hash 0x0c591f26891d6443cf08c5be3584c1e6ae10a4c2f07c5c53218741e9755fb9cd
```
shows always
```
Starting to generate deterministic contract address with 1 threads...
```
whereas on master is 16 for me.

Probably this one should also follow the config
https://github.com/foundry-rs/foundry/blob/a65a5b1445ba7ec9b10baf7ecb28f7a65bbb13ce/crates/cast/src/lib.rs#L2062-L2063

zerosnacks commented:
> impl looks good! there's a regression re the way rayon thread pool is spawned though, a command like

Thanks, good catch! Fixed

> I think we should keep the current behavior, that is if `--jobs` flag not passed then we use rayon default (the number of CPUs available) instead single thread.

I had implemented it this way to avoid spawning the global thread pool as the vast majority of commands are not multi-threaded, cc @DaniPopes should we spawn by default?

DaniPopes commented on crates/cast/bin/cmd/create2.rs:
```diff
@@ -197,19 +193,21 @@ impl Create2Args {
 
         // Loops through all possible salts in parallel until a result is found.
         // Each thread iterates over `(i..).step_by(n_threads)`.
-        for i in 0..n_threads {
```
Shouldn't change this, rayon has some overhead here

DaniPopes commented on crates/cli/src/opts/global.rs:
```diff
@@ -0,0 +1,115 @@
+            if let Some(num_threads) = num_threads {
+                return Some(num_threads.min(2));
+            }
+        }
```
This is actually the opposite, it caps at 2; this was just to reduce pressure on tests in cast create2, should keep it there

zerosnacks commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -275,13 +271,6 @@ impl TestArgs {
         // Merge all configs.
         let (mut config, mut evm_opts) = self.load_config_and_evm_opts_emit_warnings()?;
 
-        // Set number of max threads to execute tests.
-        // If not specified then the number of threads determined by rayon will be used.
-        if let Some(test_threads) = config.threads {
-            trace!(target: "forge::test", "execute tests with {} max threads", test_threads);
-            rayon::ThreadPoolBuilder::new().num_threads(test_threads).build_global()?;
-        }
```
we should indeed spawn the default threads here, this is incorrect

grandizzy commented:
wonder how (if) this would also affect the threadpools spawned in deps (and if we want to or not) like compilers

https://github.com/foundry-rs/compilers/blob/034ecd611eef030217c4b363a794226cd50b4b9e/crates/compilers/src/compile/project.rs#L529-L530

zerosnacks commented:
Decided to remove the global thread pool for now and renamed https://github.com/foundry-rs/foundry/issues/8408 as a follow-up to add it

grandizzy reviewed (approved):
lgtm!
